compiler/rustc_codegen_llvm/src/consts.rs

```diff
@@ -525,11 +525,41 @@ impl<'ll> StaticMethods for CodegenCx<'ll, '_> {
 
                 // The semantics of #[used] in Rust only require the symbol to make it into the
                 // object file. It is explicitly allowed for the linker to strip the symbol if it
-                // is dead. As such, use llvm.compiler.used instead of llvm.used.
+                // is dead, which means we are allowed use `llvm.compiler.used` instead of
+                // `llvm.used` here.
+                //
                 // Additionally, https://reviews.llvm.org/D97448 in LLVM 13 started emitting unique
                 // sections with SHF_GNU_RETAIN flag for llvm.used symbols, which may trigger bugs
-                // in some versions of the gold linker.
-                self.add_compiler_used_global(g);
+                // in the handling of `.init_array` (the static constructor list) in versions of
+                // the gold linker (prior to the one released with binutils 2.36).
+                //
+                // However, unconditional use of `llvm.compiler.used` caused a nontrivial amount of
+                // ecosystem breakage, especially on Mach-O targets. To resolve this, we compile it
+                // as llvm.compiler.used on ELF targets and llvm.used elsewhere, which should be
+                // equivalent to how we compiled `#[used]` before LLVM 13, as `llvm.used` and
+                // `llvm.compiler.used` were treated the same on ELF targets prior in earlier LLVM
+                // versions (additionally, it seems to be how Clang handles `__attribute__((used))`,
+                // perhaps for similar compatibility-motivated reasons).
+                //
+                // See https://github.com/rust-lang/rust/issues/47384#issuecomment-1019080146 and
+                // following comments for some discussion of this.
+                //
+                // The final wrinkle is it's not really clear how to tell if we're going to output
+                // ELF, so it's been approximated as "not like wasm, osx, or windows", which is
+                // not exactly correct, but is pretty close and hopefully handles all the platforms
+                // platforms where old versions of `ld.gold` are likely to show up.
+                //
+                // All this is subject to change in the future. Which is a good thing, because this
+                // probably should be firmed up somehow!
+                let seems_like_elf = !(self.tcx.sess.target.is_like_osx
+                    || self.tcx.sess.target.is_like_windows
+                    || self.tcx.sess.target.is_like_wasm);
+
+                if seems_like_elf {
+                    self.add_compiler_used_global(g);
+                } else {
+                    self.add_used_global(g);
+                }
```

Comment:
As there are separate USED and USED_LINKER flags now, this should be handled at a higher level (where used is defaulted to either used(compiler) or used(linker)).

Reply:
Damn. I won’t have time to look into this for a week ir two, sadly

Reply:
okay, based on this dialogue, I'm going to revise the state of this PR to S-waiting-on-author

@rustbot label: -S-waiting-on-review +S-waiting-on-author

Reply:
I moved this but I don't think this is a good place for it. It seems strange to have these checks in `rustc_typeck`, since I'd expect it not to care so much about target details like this.

That said, I'm not familiar enough with the compiler to say for sure.